Next set of Phlex superform components

nimmolo opened the pull request:
This PR converts four observation-related forms from ERB partials to Phlex components and enhances the component test infrastructure.

###  New Phlex Components (4)

  1. ExternalLinkForm - Create/edit external links on observations
  2. HerbariumRecordForm - Create/edit herbarium records
  3. CollectionNumberForm - Create/edit collection numbers
  4. SequenceForm - Create/edit DNA sequences

###  Transitioning to Phlex

  - `AutocompleterField` - Reusable Phlex autocompleter component
  - `modal_form` can render Form components or form partials

###  Test Coverage

  All four new forms have comprehensive test coverage:

  - Component tests (4 new files, 46 tests) - Test form rendering and attributes using `assert_html`
    - `assert_html` helper improvements: avoids HTML dumps in test failures
  - Controller tests - Updated to pass proper form_locals to components
    - Standard CRUD operations, turbo variants for modal rendering
  - System tests - Integration tests for turbo forms in `observation_show_system_test.rb`

nimmolo committed:
Collection number and sequence forms with modal_form

nimmolo committed:
more forms, autocompleter,  tests

nimmolo committed:
refactor form tests to assert_html

nimmolo committed:
Fixes for kwargs to modal form

nimmolo committed:
Update autocompleter_field.rb

nimmolo committed:
rubocop

nimmolo committed:
More assertions

nimmolo committed:
rubocop and test helper refactor

nimmolo committed:
Merge branch 'main' into nimmo-phlex-superform3

nimmolo committed:
Autocompleter improvements

nimmolo committed:
autocompleter and text field extras

nimmolo committed:
Remove buttons

JoeCohen commented:
Add External Link modal form shows numbers instead of site names in the External Site drop-down
<img width="470" height="209" alt="Screenshot 2025-11-23 at 1 41 25 PM" src="https://github.com/user-attachments/assets/06f0d885-2fa8-4042-af38-8440d34b01b4" />

JoeCohen reviewed (commented):
Thank you for moving forward with Phlex.  I'm looking forward to trying it out once you're done.
There's a RuboCop comment and some CoPilot suggestions.
I've also left some comments in the Comment thread.

nimmolo committed:
Fixes

nimmolo commented on app/components/external_link_form.rb:
```diff
@@ -0,0 +1,110 @@
+
+# Form for creating or editing external links for observations
+class Components::ExternalLinkForm < Components::ApplicationForm
+  def initialize(model, **kwargs)
```
This is a common Ruby strategy for avoiding long explicit keyword parameter lists, which our Rubocop doesn't like (and neither do I). Using the splat operator in reverse just scoops every kwarg into a single hash, and then you can destructure it.

The `initialize` method in `ApplicationForm` (called by `super` here) requires certain kwargs, so doing it this way (with `kwargs.delete`) allows this form to take the ones it needs without messing up the common, inherited form initialization -- it forwards the remaining kwargs to `super` without needing to write `super(*, **, &)`.

nimmolo committed:
Fix external sites controller

nimmolo committed:
add coverage

nimmolo committed:
Fix new Phlex email form params in controllers

Most of our forms concern db models. But some of our older forms just send params.

Phlex superforms all need to use a form object - whether it's the database record or a PORO (using ActiveModel), superform wants to validate the object you're creating/updating when you send the form, as it should - this is a sturdier code pattern than validating params individually. The params become attributes of that form object.

So the Phlex conversion has involved creating objects for object-less forms. But claude and i forgot to update the controllers to look for the params nested under these objects. This PR fixes that.

This "form-object" thing is incidentally the same pattern we now use on Search Forms.

nimmolo committed:
rubocop

nimmolo committed:
Remove param fallbacks and update tests

nimmolo committed:
Make it a controller test
